### PR TITLE
fix(chainsync): prune seenHeaders during normal operation

### LINKED ISSUE #2269: chainsync: prune seenHeaders during normal operation
## Summary

`chainsync.State.seenHeaders` can still grow without bound.

## Audit Details

Finding ID: M4, MEDIUM network/memory finding.

Affected file: `chainsync/chainsync.go`.

The audit found that `seenHeaders` tracks observed header hashes per slot and has pruning helpers, but pruning was not wired into normal runtime paths. Current code still defines `PruneSeenHeaders`, `ClearSeenHeaders`, and `ClearSeenHeadersFrom`, but references show pruning is tested rather than called during steady progress.

Impact: memory grows slowly over long-running nodes as more slots are observed. The audit estimated this as a slow leak, not an immediate OOM.

## What Needs to Happen

1. Prune `seenHeaders` based on the current stability window or another documented retention policy.
2. Call pruning from a normal progress path.
3. Keep enough history for duplicate/fork detection semantics.
4. Add tests proving old entries are pruned while recent duplicate detection still works.
5. Consider a metric for current `seenHeaders` size.

## Acceptance Criteria

- Long-running ChainSync progress prunes old header observations.
- Duplicate detection remains covered by tests after pruning.

### PR DESCRIPTION
**Pull request by cryptodj413**
Closes #2269 

---
## Summary by cubic
Automatically prune the `chainsync` header dedup cache (`seenHeaders`) during normal progress to keep memory bounded without weakening duplicate/fork detection. Adds a retention override and a `prometheus` gauge to track cache size.

- **Bug Fixes**
  - Prune `seenHeaders` based on the ledger stability window (fallback: 2500*20 slots), triggered on `UpdateClientTip`.
  - Rate-limit pruning scans to every 1000 slots of progress.
  - Tests cover pruning, fork detection on recent slots, and bounded cache growth.

- **New Features**
  - Config: `SeenHeadersRetention` to override the retention window (slots).
  - Metrics: `dingo_chainsync_seen_headers` gauge via optional `PromRegistry` (`promauto`), wired in `node.go`.
  - Utility: `SeenHeadersLen()` for tests/metrics.

<sup>Written for commit 1d78a8f47b614cc1c13ccf672fc777fc652c1929. Summary will update on new commits. <a href="https://cubic.dev/pr/blinklabs-io/dingo/pull/2428?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Prometheus metrics added to report header-deduplication cache size and behavior
  * Configurable cache retention window with automatic pruning to keep memory bounded
  * Option to register these metrics with the node's Prometheus registry

* **Tests**
  * New tests verifying pruning on progress, fork detection after pruning, and long-run bounded cache behavior

[Review Change Stack](https://app.coderabbit.ai/change-stack/blinklabs-io/dingo/pull/2428?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: chainsync/chainsync.go [hunk 1]
```diff
@@ -205,6 +239,14 @@ func NewStateWithConfig(
 			map[ouroboros.ConnectionId]*observedHeaderChain,
 		),
 	}
+	// promauto.With(nil) returns a factory that creates the gauge without
+	// registering it, so this is safe when PromRegistry is unset.
+	s.seenHeadersGauge = promauto.With(cfg.PromRegistry).NewGauge(
+		prometheus.GaugeOpts{
+			Name: "cardano_node_metrics_chainsync_seenHeaders_int",
```
Comment (wolf31o2) [inline chainsync/chainsync.go]: This is not a `cardano_node` metric.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
Rename the metric to `dingo`

**Review by wolf31o2 — APPROVED** [review]
